src/source/src/parser/mod.rs

```diff
@@ -234,6 +238,40 @@ impl SourceStreamChunkRowWriter<'_> {
         self.do_action::<OpActionInsert>(f)
     }
 
+    /// For other op like 'insert', 'update', 'delete', we will leave the hollow for the meta column
+    /// builder. e.g after insert
+    /// `data_budiler` = [1], `meta_column_builder` = [], `op` = [insert]
+    ///
+    /// This function is used to fulfill this hollow in `meta_column_builder`.
+    /// e.g after fulfill
+    /// `data_budiler` = [1], `meta_column_builder` = [1], `op` = [insert]
+    pub fn fulfill(
+        &mut self,
+        mut f: impl FnMut(&SourceColumnDesc) -> Result<Option<Datum>>,
+    ) -> Result<WriteGuard> {
+        let mut modify_col = vec![];
+
+        self.descs
+            .iter()
+            .zip_eq(self.builders.iter_mut())
+            .enumerate()
+            .try_for_each(|(idx, (desc, builder))| -> Result<()> {
+                if let Some(output) = f(desc)? {
+                    builder.append_datum(output);
+                    modify_col.push(idx);
+                }
+                Ok(())
+            })
+            .inspect_err(|e| {
+                tracing::warn!("failed to parse source data: {}", e);
```

Comment:
I'm wondering do we really need a `Result` here? If we failed, the whole row should be reverted, not just this builder for meta column.